### PR TITLE
networkd: add basic Varlink interface

### PR DESCRIPTION
**Pull request by poettering**
Let's get networkd onto Varlink. This only adds the most basic of operations.

I'd love to see networkd do Varlink for all its basic operations so that networkctl can use that, and work correctly before D-Bus is up. Right now, many of networkctls calls simply don't work before D-Bus, and I'd like to see that improved.

(This also prepares ground for when we add netns support to networkd, i.e. with one instance per netns, as we can easily bind a separate varlink socket for each, which is much more natural than doing something similar on D-Bus)

### CODE REVIEW

**Review by mdlayher — APPROVED** [review]
LGTM from a user perspective! I've been poking with varlink for a couple of days and am particularly excited to see it come to networkd.

I was able to implement a quick patch for networkctl to use GetNamespaceId and will submit that as a PR once this is merged: https://github.com/mdlayher/systemd/commit/fd12417fb7707645b2c0b8818de2274f730451a1

**Comment by poettering** [comment]
@yuwata any chance you can do a quick review of this one?

File: src/network/meson.build [hunk 1]
```diff
@@ -56,8 +56,9 @@ sources = files(
         'networkd-link.c',
         'networkd-lldp-rx.c',
         'networkd-lldp-tx.c',
-        'networkd-manager-bus.c',
         'networkd-manager.c',
+        'networkd-manager-bus.c',
```
Comment (yuwata) [inline src/network/meson.build:60]: Not a matter, but here you sort `-` after `.`, but in below you use the inverse order.

File: src/network/networkd-manager-varlink.c [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <unistd.h>
+
+#include "networkd-manager-varlink.h"
+#include "varlink-io.systemd.Network.h"
+#include "varlink.h"
```
Comment (yuwata) [inline src/network/networkd-manager-varlink.c]: Here.

File: src/network/networkd-manager-varlink.c [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        if (json_variant_elements(parameters) > 0)
+                return varlink_error_invalid_parameter(link, parameters);
+
+        return varlink_replyb(link,
+                              JSON_BUILD_OBJECT(
+                                              JSON_BUILD_PAIR_STRING("AddressState", link_address_state_to_string(m->address_state)),
+                                              JSON_BUILD_PAIR_STRING("IPv4AddressState", link_address_state_to_string(m->ipv4_address_state)),
+                                              JSON_BUILD_PAIR_STRING("IPv6AddressState", link_address_state_to_string(m->ipv6_address_state)),
+                                              JSON_BUILD_PAIR_STRING("CarrierState", link_carrier_state_to_string(m->carrier_state)),
+                                              JSON_BUILD_PAIR_STRING("OnlineState", link_online_state_to_string(m->online_state)),
```
Comment (yuwata) [inline src/network/networkd-manager-varlink.c]: Online state can be NULL. I think it is better to conditionalize the field, or accept a null or empty string.

**Review by yuwata — APPROVED** [review]
Looks good to me. One simple request, and nitpicky comments about sorting.

**Pull request merged**